src/main/kotlin/com/team4099/robot2023/subsystems/elevator/Elevator.kt

```diff
@@ -0,0 +1,356 @@
+                }
+                else -> {}
+            }
+            field = value
+        }
+
+    var elevatorPositionTarget = 0.0.inches
+        private set
+    var elevatorVelocityTarget = 0.0.inches.perSecond
+        private set
+    var elevatorVoltageTarget = 0.0.volts
+        private set
+
+    private var lastRequestedPosition = -9999.inches
+    private var lastRequestedVelocity = -9999.inches.perSecond
+    private var lastRequestedVoltage = -9999.volts
+
+    private var timeProfileGeneratedAt = Clock.fpgaTime
+
+    private var lastHomingStatorCurrentTripTime = Clock.fpgaTime
+
+    // Trapezoidal Profile Constraints
+    private var elevatorConstraints: TrapezoidProfile.Constraints<Meter> =
+        TrapezoidProfile.Constraints(ElevatorConstants.MAX_VELOCITY, ElevatorConstants.MAX_ACCELERATION)
+    private var elevatorSetpoint: TrapezoidProfile.State<Meter> =
+        TrapezoidProfile.State(inputs.elevatorPosition, inputs.elevatorVelocity)
+    private var elevatorProfile =
+        TrapezoidProfile(
+            elevatorConstraints,
+            TrapezoidProfile.State(-9999.inches, -9999.inches.perSecond),
+            TrapezoidProfile.State(-9999.inches, -9999.inches.perSecond)
+        )
+
+    val isAtTargetedPosition: Boolean
+        get() =
+            (currentRequest is ElevatorRequest.TargetingPosition &&
+                    elevatorProfile.isFinished(Clock.fpgaTime - timeProfileGeneratedAt) &&
+                    (inputs.elevatorPosition - elevatorPositionTarget).absoluteValue <=
+                    ElevatorConstants.ELEVATOR_TOLERANCE) ||
+                    (TunableElevatorHeights.enableElevator.get() != 1.0)
+
+    val canContinueSafely: Boolean
+        get() =
+            currentRequest is ElevatorRequest.TargetingPosition && (
+                    ((inputs.elevatorPosition - elevatorPositionTarget).absoluteValue <= 5.inches) ||
+                            elevatorProfile.isFinished(Clock.fpgaTime - timeProfileGeneratedAt)
+                    ) && lastRequestedPosition == elevatorPositionTarget
+
+
+    init {
+        TunableElevatorHeights
+
+        // Initializes PID constants and changes FF depending on if sim or real
+        if(RobotBase.isReal()) {
+            isHomed = false
+
+            kP.initDefault(ElevatorConstants.REAL_KP)
+            kI.initDefault(ElevatorConstants.REAL_KI)
+            kD.initDefault(ElevatorConstants.REAL_KD)
+        } else {
+            isHomed = true
+
+            kP.initDefault(ElevatorConstants.SIM_KP)
+            kI.initDefault(ElevatorConstants.SIM_KI)
+            kD.initDefault(ElevatorConstants.SIM_KD)
+        }
+
+        elevatorFeedforward = ElevatorFeedforward(
+            ElevatorConstants.ELEVATOR_KS,
+            ElevatorConstants.ELEVATOR_KG,
+            ElevatorConstants.ELEVATOR_KV,
+            ElevatorConstants.ELEVATOR_KA
+        )
+
+        io.configPID(kP.get(), kI.get(), kD.get())
```

Comment:
im not sure if this will work. `init` runs on object instantiation, but the communication interface to motors might not be up by then, worst case stick it in a function and run on `robotInit`